src/Interpreters/ExpressionAnalyzer.cpp

```diff
@@ -1337,10 +1343,12 @@ ExpressionAnalysisResult::ExpressionAnalysisResult(
 
         if (query_analyzer.hasTableJoin())
         {
-            query_analyzer.appendJoinLeftKeys(chain, only_types || !first_stage);
+            Block left_block_sample = source_header;
+
+            query_analyzer.appendJoinLeftKeys(chain, only_types || !first_stage, left_block_sample);
```

Comment:
Did you consider adding the necessary casts to the right table here, using the ExpressionTransform that happens with it before joins? And then using this casted column as the right-side key. Probably you'll need much less special logic in other places then...

Reply:
I thought about this but I not sure that it doesn't break types in result table. We need to restore types in joined table. It's clearer to me how to perform conversion inside join implementation without affecting pipeline, but I realize that transform may be better...